src/main/scala/diplomacy/Nodes.scala

```diff
@@ -633,11 +709,39 @@ class SourceNode[D, U, EO, EI, B <: Data](imp: NodeImp[D, U, EO, EI, B])(po: Seq
 {
   override def description = "source"
   protected[diplomacy] def resolveStar(iKnown: Int, oKnown: Int, iStars: Int, oStars: Int): (Int, Int) = {
-    require (oStars <= 1, s"$context appears right of a :=* ${oStars} times; at most once is allowed")
-    require (iStars == 0, s"$context cannot appear left of a :*=")
-    require (iKnown == 0, s"$context cannot appear left of a :=")
-    require (po.size == oKnown || oStars == 1, s"$context has only ${oKnown} outputs connected out of ${po.size}")
-    require (po.size >= oKnown, s"$context has ${oKnown} outputs out of ${po.size}; cannot assign ${po.size - oKnown} edges to resolve :=*")
+    def resolveStarInfo: String =
+      s"""$context
+         |$bindingInfo
+         |inward nodes already known: $iKnown
+         |outward nodes already known: $oKnown
+         |query by nodes other nodes: $iStars
+         |query to other nodes: $oStars
+         |${po.size} parameterized sinks: [${po.map(_.toString).mkString(",")}]
+         |""".stripMargin
+    require(oStars <= 1,
+      s"""$resolveStarInfo
+         |This node appears right of a :*= $oStars times; at most once is allowed.
+         |""".stripMargin)
+    require(iStars == 0,
+      s"""$resolveStarInfo
+         |This node cannot appear left of a :*=
+         |""".stripMargin)
+    require(iKnown == 0,
+      s"""$resolveStarInfo
+         |This node cannot appear left of a :=
+         |""".stripMargin)
+    if (oStars == 0)
+      require(po.size == oKnown,
+        s"""$resolveStarInfo
+           |$oKnown outputs were connected but ${po.size} sources were specified to the node constructor.
```

Comment:
`|$oKnown outward bindings were connected but ${po.size} sources were specified to the node constructor.`